app/connection/robust-websocket.js

```diff
@@ -0,0 +1,246 @@
+    }
+
+    ['bufferedAmount', 'url', 'readyState', 'protocol', 'extensions'].forEach(function(readOnlyProp) {
+      Object.defineProperty(self, readOnlyProp, {
+        get: function() { return realWs[readOnlyProp] }
+      })
+    })
+
+    function clearPendingReconnectIfNeeded() {
+      if (pendingReconnect) {
+        clearTimeout(pendingReconnect)
+        pendingReconnect = null
+      }
+    }
+
+    var ononline = function(event) {
+      if (reconnectWhenOnlineAgain) {
+        clearPendingReconnectIfNeeded()
+        reconnect(event)
+      }
+    },
+    onoffline = function() {
+      reconnectWhenOnlineAgain = true
+      realWs.close(1000)
+    },
+    connectivityEventsAttached = false
+
+    function detachConnectivityEvents() {
+      if (connectivityEventsAttached) {
+        global.removeEventListener('online', ononline)
+        global.removeEventListener('offline', onoffline)
+        connectivityEventsAttached = false
+      }
+    }
+
+    function attachConnectivityEvents() {
+      if (!connectivityEventsAttached) {
+        global.addEventListener('online', ononline)
+        global.addEventListener('offline', onoffline)
+        connectivityEventsAttached = true
+      }
+    }
+
+    self.send = function() {
+      return realWs.send.apply(realWs, arguments)
+    }
+
+    self.close = function(code, reason) {
+      if (typeof code !== 'number') {
+        reason = code
+        code = 1000
+      }
+
+      clearPendingReconnectIfNeeded()
+      reconnectWhenOnlineAgain = false
+      explicitlyClosed = true
+      detachConnectivityEvents()
+
+      return realWs.close(code, reason)
+    }
+
+    self.open = function() {
+      if (realWs.readyState !== WebSocket.OPEN && realWs.readyState !== WebSocket.CONNECTING) {
+        clearPendingReconnectIfNeeded()
+        reconnectWhenOnlineAgain = false
+        explicitlyClosed = false
+
+        newWebSocket()
+      }
+    }
+
+    function reconnect(event) {
+      if ((!opts.shouldReconnect.handle1000 && event.code === 1000) || explicitlyClosed) {
+        attempts = 0
+        return
+      }
+      if (navigator.onLine === false) {
+        reconnectWhenOnlineAgain = true
+        return
+      }
+
+      var delay = opts.shouldReconnect(event, self)
+      if (typeof delay === 'number') {
+        pendingReconnect = setTimeout(newWebSocket, delay)
+      }
+    }
+
+    Object.defineProperty(self, 'listeners', {
+      value: {
+        open: [function(event) {
+          if (connectTimeout) {
+            clearTimeout(connectTimeout)
+            connectTimeout = null
+          }
+          event.reconnects = ++reconnects
+          event.attempts = attempts
+          attempts = 0
+          reconnectWhenOnlineAgain = false
+        }],
+        close: [function(event){
+          setTimeout(() => {

```

Comment:
@brucebinary here when the mobile phone wifi goes offline, the ws connection closes before browser detect the connection change so `navigator.online` is still true when it tries to reconnect.